### DISCUSSION TITLE
bad experience when brew upgraded dependencies

### DISCUSSION THREAD

**Discussion by jsoref** [jj-vcs/jj#1570] [General]
```rust
Last login: Tue May  2 13:46:13 on ttys217
dyld[83152]: Library not loaded: /opt/homebrew/opt/libgit2/lib/libgit2.1.5.dylib
  Referenced from: <4262F695-957A-386A-9408-5479E5DEC4BC> /Users/jsoref/.cargo/bin/jj
  Reason: tried: '/opt/homebrew/opt/libgit2/lib/libgit2.1.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/libgit2/lib/libgit2.1.5.dylib' (no such file), '/opt/homebrew/opt/libgit2/lib/libgit2.1.5.dylib' (no such file), '/usr/local/lib/libgit2.1.5.dylib' (no such file), '/usr/lib/libgit2.1.5.dylib' (no such file, not in dyld cache), '/opt/homebrew/Cellar/libgit2/1.6.4/lib/libgit2.1.5.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/libgit2/1.6.4/lib/libgit2.1.5.dylib' (no such file), '/opt/homebrew/Cellar/libgit2/1.6.4/lib/libgit2.1.5.dylib' (no such file), '/usr/local/lib/libgit2.1.5.dylib' (no such file), '/usr/lib/libgit2.1.5.dylib' (no such file, not in dyld cache)
```

`~/.zshrc`:
```sh
if type jj &>/dev/null; then
  autoload -U compinit
  compinit
  source <(jj debug completion --zsh | sed '$d')  # remove the last line
  compdef _jj jj
fi
```

This happened in each window as I opened new terminals.

Once I reinstalled `jj` w/ the latest version (`cargo install --git https://github.com/martinvonz/jj.git --locked --bin jj jujutsu`), I got this:

```rust
Last login: Tue May  2 13:48:26 on ttys218
error: unrecognized subcommand 'completion'

  note: subcommand 'operation' exists
  note: to pass 'completion' as a value, use 'jj debug -- completion'

Usage: jj debug [OPTIONS] <COMMAND>

For more information, try '--help'.
```

Which is probably slightly better.

If there's going to be a `zsh` or similar instruction recommended to be added to a shell start script, going forward it should probably handle the cases where the command goes away or where the libraries are missing so that it isn't so noisy and unhelpful.

**Comment by martinvonz** [comment]
You might have figured it out by now, but `jj debug completion` is now called `jj util completion`. Also, the script it generates is different in a way that makes the way you source it different. See the updated instructions at https://github.com/martinvonz/jj#zsh.

We could deal with the command rename by keeping `jj debug completion` as a hidden alias. Maybe the project has enough users  by now that we should have. It's just hard to tell how many users there are who are not keeping up the project on Discord and GitHub PRs, so thanks for letting us know.

The other change is harder to make compatible. I suppose we could have added a `--new-zsh-variant` to get the new behavior and then made that the default only a year or two later? I think the project is still too small to need that level of compatibility.

**Reply by jsoref** [reply]
So, I'd start by not using `source <(jj debug completion --zsh...)`

Instead, do something like:
```zsh
jj_completion_script=$(mktemp)
jj_completion_errors=$(mktemp)
if jj debug completion --zsh > "$jj_completion_script" 2> "$jj_completion_errors"; then
  source | <(sed '$d' $"jj_completion_script")
fi
```

**Reply by jsoref** [reply]
Tentatively, I'm going to try:
```zsh
if type jj &>/dev/null && jj -V &>/dev/null; then
  autoload -U compinit
  compinit
  source <(jj util completion --zsh)
fi
```

Which I hope will guard against the next time that jj's library dependencies break. (It won't cover the case of `jj util completion --zsh` crashing, but...)
